### ISSUE TITLE
Tye integration with ARM

### ISSUE THREAD

**Issue by david-t-daniel** [dotnet/tye#854]
### What should we add or change to make your life better?
I would love to see tye be able to deploy to Azure by compiling its settings to an ARM AST like [Bicep](https://github.com/Azure/bicep) and generating the ARM templates for deployment.  An example would be when running locally it uses a recipe like redis, azure function or .Net program but if you deployed to Azure it would generate the ARM templates for redis, azure functions and an app service and deploys them to your Azure instance.

### Why is this important to you?
I think it will help ease deployment and local development for stuff that may not run in Kubernetes like Azure Storage Emulator or Cosmos DB Emulator.

**Comment by jongio** [comment]
This is an interesting suggestion and speaks to how complex/overcomplicated simple scenarios become because of all of the knobs we give users.

Ideally we publish a set of recipes that enable users to quickly get up and running that include Tye and Bicep integration.  I'm not sure combining or generating Bicep is the way to go, but definitely something worth exploring....more thoughts below.


I personally like to separate infrastructure as code and platform as code, so I can deploy the infra once and only update as needed and version it independently of my platform code.

I'm using Bicep as IaC and Tye as PaC here:  https://github.com/jongio/memealyzer#quickstart

I was thinking through this request and considering what the combined of those two things would look like.

With IaC, you are required to set a lot of parameters and settings that don't necessarily apply to running the app.  For example, you have an underlying storage account, but don't define that in Tye, or consume an API from a SaaS service, that you deploy, i.e. Cognitive Services - but you don't define that in Tye.

I think for a simple recipe it may work, but we'd have a lot of defaulting and would likely need to jump deep into tweaking.

Let's take the example of running and deploying an azure function:

Tye:
```yaml
  - name: memealyzernetfunction
    azureFunction: ${ROOT}/src/net/Services/Functions/
    bindings:
      - port: 3080
        protocol: http
    tags:
      - local
```


Bicep:
```
resource function 'Microsoft.Web/sites@2020-06-01' = {
  name: '${basename}function'
  location: location
  kind: 'functionapp'
  identity: {
    type: 'SystemAssigned'
  }
  properties: {
    serverFarmId: plan.id
    siteConfig: {
      alwaysOn: true
      cors: {
        allowedOrigins: [
          '*'
        ]
        supportCredentials: false
      }
      ftpsState: 'FtpsOnly'
    }
  }
}

resource function_app_settings 'Microsoft.Web/sites/config@2018-11-01' = {
  name: '${basename}function/appsettings'
  properties: {
    'AZURE_KEYVAULT_ENDPOINT': key_vault.properties.vaultUri
    'AzureWebJobsStorage': 'DefaultEndpointsProtocol=https;AccountName=${storage.name};EndpointSuffix=${environment().suffixes.storage};AccountKey=${listKeys(storage.id, storage.apiVersion).keys[0].value}'
    'APPINSIGHTS_INSTRUMENTATIONKEY': logging.properties.InstrumentationKey
    'FUNCTIONS_WORKER_RUNTIME': 'dotnet'
    'FUNCTIONS_EXTENSION_VERSION': '~3'
    'WEBSITES_ENABLE_APP_SERVICE_STORAGE': 'false'
    'AZURE_CLIENT_SYNC_QUEUE_NAME': 'sync'
    'AZURE_STORAGE_CONNECTION_STRING_SECRET_NAME': 'StorageConnectionString'
    'AZURE_SIGNALR_CONNECTION_STRING_SECRET_NAME': 'SignalRConnectionString'
    'WEBSITE_RUN_FROM_PACKAGE': ''
  }
}

resource key_vault 'Microsoft.KeyVault/vaults@2019-09-01' = {
  name: '${basename}kv'
  location: location
  properties: {
    tenantId: subscription().tenantId
    sku: {
      family: 'A'
      name: 'standard'
    }
    accessPolicies: [
      {
        objectId: cli_user_id
        permissions: {
          secrets: secrets
        }
        tenantId: subscription().tenantId
      }
      {
        objectId: aks.properties.identityProfile.kubeletidentity.objectId
        permissions: {
          secrets: secrets
        }
        tenantId: subscription().tenantId
      }
      {
        objectId: aks.identity.principalId
        permissions: {
          secrets: secrets
        }
        tenantId: subscription().tenantId
      }
      {
        objectId: function.identity.principalId
        permissions: {
          secrets: secrets
        }
        tenantId: subscription().tenantId
      }
    ]
  }
}

resource storage 'Microsoft.Storage/storageAccounts@2019-06-01' = {
  name: '${basename}storage'
  location: location
  kind: 'StorageV2'
  sku: {
    name: 'Standard_LRS'
  }
  properties: {}
}
```


If you are using KeyVault to store secrets, then you'd also need to deploy KeyVault, and then create an access policy for the Managed Identity of the Azure Function to access those secrets to talk to storage etc.


@jkotalik @alex-frankel

**Comment by david-t-daniel** [comment]
You are right and in my mind I was not thinking that all my config would be in tye, or bicep because there are things that fall outside the scope of both like region deployments and HA settings.  My thinking was more about how tye resources would map to ARM rather than just Kubernetes.  I was bringing up bicep more for the AST that allows them to take the bicep configuration and compile to ARM templates but it also allows them to take ARM templates and create a bicep file.  If I could have ARM templates that have all the extra values that tye does not care about but somehow map tie to those resources so it could know what recipes to start when I run locally and where to deploy when I tell it to then I am happy to have multiple tools work together.

**Comment by jongio** [comment]
Sounds good.  I would also like to see tye work as a general purpose Azure deployment tool, not just k8s.  Right now, I have a deploy.sh script that does the Tye deployment but it also does the function deployment because that is hosted in a separate azure resource.  We have an issue to support function deployment, but it would make sense to support deployment of any app service, not just functions.

So, in tye.yaml, maybe we have a property that maps the tye service to the azure resource it should be deployed to.

```yaml
  - name: memealyzernetfunction
    azureFunction: ${ROOT}/src/net/Services/Functions/
    azureResourceName: memealyzerazurefunction1
    bindings:
      - port: 3080
        protocol: http
    tags:
      - local
```

And then when tye deploy is called, it maps that service to the name and deploys it using the tools for that service, zip deploy, function CLI, etc.

Related: #731

**Comment by davidfowl** [comment]
I think in the long run, we'd need tye to have an extensible provider model much like we have right now with extensions. Though they are hardcoded into this repo, the idea is that extensions are able to observe and mutate the runtime object model. This is how the dapr extension allows sidecars etc.

For deployment, it's more of a compiler model where there's a model that can be converted into whatever IaC tooling you want to use. Today we hard code k8s but it's conceivable to support other systems.
